File: src/main/java/io/testomat/api/common/ResponseDecorator.java
package io.testomat.api.common;

import io.restassured.response.Response;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;
import org.assertj.core.api.Assertions;

@NoArgsConstructor
public class ResponseDecorator<T> {

    private Response targetResponse;
    private int expectedDefaultStatusCode;
    private Class<T> targetClass;

    public ResponseDecorator(Response targetResponse, int expectedDefaultStatusCode, Class<T> targetClass) {
        this.targetResponse = targetResponse;
        this.expectedDefaultStatusCode = expectedDefaultStatusCode;
        this.targetClass = targetClass;
    }

    public ResponseDecorator assertStatusCode(int statusCode) {
        Assertions.assertThat(targetResponse.statusCode()).withFailMessage(
                String.format(
                        "Expected status code %s, but was %s \nresponse body was %s",
                        statusCode,
                        targetResponse.statusCode(),
                        targetResponse.body().asPrettyString()
                )
        ).isEqualTo(statusCode);
        return this;
    }

    @SneakyThrows
    public <T> T toObject() {
        assertStatusCode(expectedDefaultStatusCode);
        try {
            return (T) as(this.targetClass);
        } catch (Exception e) {
            throw BindingExceptionHandler.catchResponseException(e, targetClass);
        }
    }

    public <T> T as(Class<T> targetClass) {
        return targetResponse.as(targetClass);
    }

    public ResponseDecorator<T> request(Response targetResponse) {
        this.targetResponse = targetResponse;
        return this;
    }

    public ResponseDecorator<T> expectedDefaultStatusCode(int expectedDefaultStatusCode) {
        this.expectedDefaultStatusCode = expectedDefaultStatusCode;
        return this;
    }

    public ResponseDecorator<T> targetClass(Class<T> targetClass) {
        this.targetClass = targetClass;
        return this;
    }

}

File: src/main/kotlin/io/testomat/common/ResponseDecorator.kt
package io.testomat.common

import io.restassured.response.Response
import org.assertj.core.api.Assertions

data class ResponseDecorator<T>(
    val targetResponse: Response,
    val defaultStatusCode: HTTPStatusCodes,
    val targetObject: Class<T>
) {

    fun expectedStatusCodeIs(expectedStatusCode: HTTPStatusCodes): ResponseDecorator<T> {
        targetResponse.expectedStatusCodeIs(expectedStatusCode)
        return this
    }

    inline fun <reified T> toObject(): T {
        Assertions.assertThat(targetResponse.statusCode()).isEqualTo(defaultStatusCode)
        try {
            return this.targetResponse.toReified<T>()
        } catch (unEx: Exception) {
            throw catchKResponseException(unEx, T::class.java)
        }
    }
}

inline fun <reified T> Response.toReified(): T {
    try {
        return this.body().`as`(T::class.java)
    } catch (unEx: Exception) {
        throw catchKResponseException(unEx, T::class.java)
    }
}


enum class HTTPStatusCodes(val code: Int) {
    OK(200),
    CREATED(201),
    ACCEPTED(202),
    NO_CONTENT(204),
    BAD_REQUEST(400),
    UNAUTHORIZED(401),
    FORBIDDEN(403),
    NOT_FOUND(404),
    METHOD_NOT_ALLOWED(405),
    CONFLICT(409),
    UNAVAILABLE_FOR_LEGAL_REASONS(451),
    INTERNAL_SERVER_ERROR(500),
    BAD_GATEWAY(502),
    SERVICE_UNAVAILABLE(503),
    GATEWAY_TIMEOUT(504),
}
